src/journal/journal.c

```diff
@@ -0,0 +1,612 @@
+               return rc;
+       }
+
+       entry = malloc(sizeof(struct JournalEntry));
+       entry->classification = NULL;
+       entry->record_id = NULL;
+       entry->event_id = NULL;
+       entry->boot_id = NULL;
+
+       if (get_random_id(&record_id) != 0) {
+#ifdef DEBUG
+               fprintf(stderr, "Error: unable to generate random id\n");
+#endif
+               goto quit;
+       }
+
+       if (read_boot_id(&boot_id) != 0) {
+#ifdef DEBUG
+               fprintf(stderr, "Error: unable to read boot_id\n");
+#endif
+               goto quit;
+       }
+
+       if ((entry->classification = strdup(classification)) == NULL) {
+               goto quit;
+       }
+       if ((entry->event_id = strdup(event_id)) == NULL) {
+               goto quit;
+       }
+
+       entry->record_id = record_id;
+       entry->boot_id = boot_id;
+       entry->timestamp = timestamp;
+
+       if ((rc = save_entry(telem_journal->fptr, entry)) == 0) {
+               telem_journal->record_count = telem_journal->record_count + 1;
+#ifdef DEBUG
+               fprintf(stdout, "%d records in journal\n", telem_journal->record_count);
+#endif
+       }
+
+quit:
+       free_journal_entry(entry);
+
+       return rc;
+}
+
+/* Exported function */
+int prune_journal(struct TelemJournal *telem_journal)
+{
+
+    int rc = 1;
+    int count = 0;
+    int deviation = DEVIATION;
+
+    if ( telem_journal->record_count > (deviation + telem_journal->record_count_limit)) {
+        count = telem_journal->record_count - telem_journal->record_count_limit;
+
+        // jump to line# count
+        if ( skip_n_lines(count, telem_journal->fptr) != 0) {
+#ifdef DEBUG
+            fprintf(stderr, "Error: skipping %d journal lines\n", count);
+#endif
+            return rc;
+        }
+        // create new file with rest of file
+        if ( copy_to_tmp(telem_journal->fptr) != 0) {
+#ifdef DEBUG
+            fprintf(stderr, "Error: copying partial journal to temp journal file\n");
+#endif
+            return rc;
+        }
+        // close file handler
+        if ( fclose(telem_journal->fptr) != 0) {
+#ifdef DEBUG
+            fprintf(stderr, "Error: clossing journal file handler\n");
+#endif
+            return rc;
+        }
+        // overwrite file
+        if ( rename(JOURNAL_TMP, JOURNAL_PATH) != 0) {
+#ifdef DEBUG
+            fprintf(stderr, "Error: while overriding journal file\n");
+#endif
+            return rc;
+        }
+        // reopen file handler
+        telem_journal->fptr = fopen(JOURNAL_PATH, "a+");
+        if ( !telem_journal->fptr ) {
+#ifdef DEBUG
+            fprintf(stderr, "Error: re-opening journal file\n");
+#endif
+            return rc;
+        }
+        // update record count
+        telem_journal->record_count = (count >= telem_journal->record_count_limit) ?
+                                      telem_journal->record_count_limit :
+                                      telem_journal->record_count - count;
```

Comment:
Shouldn't it _always_ be `->record_count - count`, since count is how many lines we skipped?